File: src/solar/solar-common/attribute/SolarDamageState.ts
import {Trigger} from "@/w3ts/handles/trigger"
import AttributeUtil from "@/util/system/AttributeUtil";
import MathUtil from "@/util/math/MathUtil";
import SingletonUtil from "../util/lang/SingletonUtil";
import DamageUtil from "../util/system/DamageUtil";
import UnitStateUtil from "@/UnitStateUtil";


/**
 * 太阳伤害系统
 * 可配置config.damageEventHandlers 以自行处理伤害事件
 */

export default class SolarDamageState {

    static config: {
        /** 启用太阳基础属性词条 */
        useBaseAttribute: boolean
        /**物理增伤 对攻击伤害有效*/
        physical_damage_increased_attack: boolean
        /**全伤害减免 上限 默认0.99=99%*/
        damage_reduction_max: number
        /**
         * 伤害处理器
         * 按数组先后顺序执行
         * 默认分3个数组 以便简单区分哪些类别的在前面 哪些类别的在后面 比如吸血放到后面 比如减少抗性放到前面
         * @param event
         */
        firstDamageEventHandlers: ((event: AppDamageEvent) => void)[]
        damageEventHandlers: ((event: AppDamageEvent) => void)[]
        lastDamageEventHandlers: ((event: AppDamageEvent) => void)[]

    } = {
        useBaseAttribute: true,
        physical_damage_increased_attack: true,
        damage_reduction_max: 0.99,
        firstDamageEventHandlers: [],
        damageEventHandlers: [],
        lastDamageEventHandlers: [],
    }


    /**
     * 添加伤害处理器 会比addEventHandler正常添加的先执行
     * 将固定数值 和百分比数值分开放处理器队列  可以获得
     * (比如法术强度加法的放到这里 法术增伤的放到正常处理 这样法术强度就能受到法术增伤加成)
     * @param eventHandler
     */
    static addEventHandlerFirst(eventHandler: (event: AppDamageEvent) => void) {
        SolarDamageState.config.firstDamageEventHandlers.push(eventHandler);
    }

    /**
     * 添加伤害处理器
     * @param eventHandler
     */
    static addEventHandler(eventHandler: (event: AppDamageEvent) => void) {
        SolarDamageState.config.damageEventHandlers.push(eventHandler);
    }

    /**
     * 添加伤害处理器 会比addEventHandler正常添加的后执行
     * @param eventHandler
     */
    static addEventHandlerLast(eventHandler: (event: AppDamageEvent) => void) {
        SolarDamageState.config.lastDamageEventHandlers.push(eventHandler);
    }

    constructor(useBaseAttribute = SolarDamageState.config.useBaseAttribute) {
        if (SingletonUtil.notFirstTime(SolarDamageState)) {
            print("不能重复new SolarDamageState()")
            return;
        }
        SolarDamageState.config.useBaseAttribute = useBaseAttribute;
        if (useBaseAttribute) {
            //计算基础的属性伤害
            SolarDamageState.addEventHandler(SolarDamageState.calculateAttributeDamage);
        }
        isSolarDamageEnable = true;
        let trigger = new Trigger();
        trigger.registerAnyUnitDamagedEvent();
        trigger.addAction(this.action);
    }


    action(this: void) {
        let eventDamage = GetEventDamage()
        //判断伤害值
        if (eventDamage <= 0) {
            return
        }
        //设置变量
        let unit0 = GetTriggerUnit()
        let unit1 = GetEventDamageSource()
        //判断相关条件
        if (!IsHandle(unit1)) {
            return;
        }
        if (IsUnitAlly(unit0, GetOwningPlayer(unit1))) {
            return
        }
        let isAttack = (0 != EXGetEventDamageData(EVENT_DAMAGE_DATA_IS_ATTACK));
        if (isBigAttributeMode && isAttack) {//大属性模式下 走自己的伤害算法
            let u1Attack = GetUnitState(unit1, UnitStateDamageMax);//攻击力
            let u0Armor = GetUnitState(unit0, UnitStateArmor);//受到攻击者的 护甲
            eventDamage = u1Attack * (1 - MathUtil.armorReduction(u0Armor))
            EXSetEventDamage(eventDamage);
        }
        if (!isSolarDamageEnable) {//防止重复死循环触发伤害事件 即在本伤害事件中造成的伤害 不再进入此触发
            return;
        }

        // 需要单位用户数据（全部为实数类型 1  = 100%）

        let u0sa: AppAttribute = AttributeUtil.getUnitAttribute(unit0);
        let u1sa: AppAttribute = AttributeUtil.getUnitAttribute(unit1);


        let resultDamage = eventDamage;
        /**计算护甲穿透*/
        if (SolarDamageState.config.useBaseAttribute && isAttack && u1sa != null) {
            if (u1sa.def_pierce_p || u1sa.def_pierce) {
                //计算穿甲比例后受害单位的护甲
                let armor = UnitStateUtil.getArmor(unit0) * (1 - (u1sa.def_pierce_p || 0))
                //计算最终受害单位生效的护甲,等于扣除了穿甲比例后,在扣除固定穿甲
                armor = armor - (u1sa.def_pierce || 0)
                //计算减伤因子
                let rd = MathUtil.armorReduction(armor)
                resultDamage = Math.max(resultDamage, UnitStateUtil.getDamageMax(unit1)) * (1 - rd)
            }
        }


        //封装太阳伤害事件
        let event: AppDamageEvent = {
            baseDamage: eventDamage,
            damageType: EXGetEventDamageData(EVENT_DAMAGE_DATA_DAMAGE_TYPE),
            weaponType: EXGetEventDamageData(EVENT_DAMAGE_DATA_WEAPON_TYPE),
            unit0: unit0,
            unit1: unit1,
            u0sa: u0sa,
            u1sa: u1sa,
            //
            isAttack: isAttack,
            isRanged: (0 != EXGetEventDamageData(EVENT_DAMAGE_DATA_IS_RANGED)),
            isPhysical: DamageUtil.isEventPhysicalDamageType(),
            isCritical: false,
            resultDamage: resultDamage,
        }

        //对太阳伤害事件 执行处理
        isSolarDamageEnable = false;
        for (let damageEventHandler of SolarDamageState.config.firstDamageEventHandlers) {
            damageEventHandler(event);
            if (event.consumed) {//已处理完毕 退出后续处理逻辑
                isSolarDamageEnable = true;
                return;
            }
        }
        for (let damageEventHandler of SolarDamageState.config.damageEventHandlers) {
            damageEventHandler(event);
            if (event.consumed) {//已处理完毕 退出后续处理逻辑
                isSolarDamageEnable = true;
                return;
            }
        }
        for (let damageEventHandler of SolarDamageState.config.lastDamageEventHandlers) {
            damageEventHandler(event);
            if (event.consumed) {//已处理完毕 退出后续处理逻辑
                isSolarDamageEnable = true;
                return;
            }
        }
        isSolarDamageEnable = true;

        // 设置伤害值
        EXSetEventDamage(event.resultDamage);

    }


    /**
     * 计算太阳基础属性伤害
     * @param event
     */
    static calculateAttributeDamage(event: SolarDamageEvent): void {
        // 判断伤害类型(计算物理法术增伤与暴击伤害)
        let u0sa = event.u0sa;
        let u1sa = event.u1sa;
        if (u0sa == null && u1sa == null) {
            return
        }
        let newResultDamage = event.resultDamage;
        if (event.isAttack && u1sa?.attack_damage_increased) {
            //攻击增伤
            newResultDamage *= (u1sa.attack_damage_increased + 1)
        }

        if (event.isPhysical) {
            //物理增伤
            if (SolarDamageState.config.physical_damage_increased_attack || event.isAttack == false) {
                newResultDamage *= ((u1sa?.physical_damage_increased ?? 0) + 1)
            }

            //物理暴击
            if ((GetRandomReal(0, 1) < (u1sa?.physical_critical_chance ?? 0))) {
                //Critical damage 暴击伤害  1 = 增加100%伤害
                newResultDamage = newResultDamage * ((u1sa?.physical_critical_damage ?? 0) + 1);
                event.isCritical = true;
            }

            //物理穿透与减免
            let physical_damage_pierce = (u1sa?.physical_damage_pierce || 0) - (u0sa?.physical_damage_reduction || 0)
            newResultDamage *= (1 + physical_damage_pierce)

        } else {
            // 法术强度
            newResultDamage += (u1sa?.magic_power ?? 0)
            // 法术增伤
            newResultDamage *= ((u1sa?.magic_damage_increased ?? 0) + 1)
            // 判断暴击
            if ((GetRandomReal(0, 1) < (u1sa?.magic_critical_chance ?? 0))) {
                //Critical damage 暴击伤害
                newResultDamage = newResultDamage * ((u1sa?.magic_critical_damage ?? 0) + 1)
                event.isCritical = true;
            }

            //法术穿透与减免
            let magic_damage_pierce = (u1sa?.magic_damage_pierce || 0) - (u0sa?.magic_damage_reduction || 0)
            newResultDamage *= (1 + magic_damage_pierce)

        }
        /**
         * 根据伤害类型 增伤
         */
        let damage_increased_dt_x = "damage_increased_dt_" + event.damageType
        if (u1sa?.[damage_increased_dt_x]) {
            newResultDamage *= (1 + (u1sa[damage_increased_dt_x] as number))
        }

        /**
         * 根据武器类型 增伤
         */
        let damage_increased_wt_x = "damage_increased_wt_" + event.weaponType
        if (u1sa?.[damage_increased_wt_x]) {
            newResultDamage *= (1 + (u1sa[damage_increased_wt_x] as number))
        }

        /**
         * 最终增伤
         */
        if (u1sa?.damage_increased) {
            newResultDamage *= (1 + u1sa.damage_increased)
        }


        /**
         *  计算 受到伤害的单位 的 属性 减免伤害 u0sa
         *  伤害减免Damage Reduction
         */
        if (u0sa?.damage_reduction) {
            newResultDamage *= (1 - Math.min(u0sa.damage_reduction, SolarDamageState.config.damage_reduction_max))
        }

        /**
         * 根据伤害类型 抗性 减伤
         */
        let damage_reduction_dt_x = "damage_reduction_dt_" + event.damageType
        if (u0sa?.[damage_reduction_dt_x]) {
            newResultDamage *= (1 - (u0sa[damage_reduction_dt_x] as number))
        }
        /**
         * 根据武器类型 抗性 减伤
         */
        let damage_reduction_wt_x = "damage_reduction_wt_" + event.weaponType
        if (u0sa?.[damage_reduction_wt_x]) {
            newResultDamage *= (1 - (u0sa[damage_reduction_wt_x] as number))
        }


        /**
         *  任意伤害吸血
         */
        if (u1sa?.blood_sucking) {
            let add_hp = newResultDamage * u1sa.blood_sucking
            SetUnitState(event.unit1, UNIT_STATE_LIFE, Math.max(0, GetUnitState(event.unit1, UNIT_STATE_LIFE) + add_hp))
        }

        //
        event.resultDamage = newResultDamage
    }


}

File: scripts/solar/solar-common/attribute/SolarDamageState.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__New = ____lualib.__TS__New
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["7"] = 1,["8"] = 1,["9"] = 2,["10"] = 2,["11"] = 3,["12"] = 3,["13"] = 4,["14"] = 4,["15"] = 5,["16"] = 5,["17"] = 6,["18"] = 6,["19"] = 14,["20"] = 14,["21"] = 14,["22"] = 69,["23"] = 69,["24"] = 69,["26"] = 70,["27"] = 71,["30"] = 74,["31"] = 75,["32"] = 77,["34"] = 79,["35"] = 80,["36"] = 81,["37"] = 82,["38"] = 69,["39"] = 49,["40"] = 50,["41"] = 50,["42"] = 49,["43"] = 57,["44"] = 58,["45"] = 58,["46"] = 57,["47"] = 65,["48"] = 66,["49"] = 66,["50"] = 65,["51"] = 86,["52"] = 87,["53"] = 89,["56"] = 93,["57"] = 94,["58"] = 96,["61"] = 99,["62"] = 99,["63"] = 99,["64"] = 99,["67"] = 102,["68"] = 103,["69"] = 104,["70"] = 105,["71"] = 106,["72"] = 107,["74"] = 109,["77"] = 115,["78"] = 116,["79"] = 119,["80"] = 121,["81"] = 122,["82"] = 124,["83"] = 126,["84"] = 128,["85"] = 129,["86"] = 129,["87"] = 129,["88"] = 129,["91"] = 135,["92"] = 135,["93"] = 135,["94"] = 135,["95"] = 135,["96"] = 135,["97"] = 135,["98"] = 135,["99"] = 135,["100"] = 135,["101"] = 135,["102"] = 135,["103"] = 135,["104"] = 135,["105"] = 152,["106"] = 153,["107"] = 154,["108"] = 155,["109"] = 156,["113"] = 160,["114"] = 161,["115"] = 162,["116"] = 163,["120"] = 167,["121"] = 168,["122"] = 169,["123"] = 170,["127"] = 174,["128"] = 177,["129"] = 86,["130"] = 186,["131"] = 188,["132"] = 189,["133"] = 190,["136"] = 193,["137"] = 194,["138"] = 196,["140"] = 199,["141"] = 201,["142"] = 202,["144"] = 206,["145"] = 208,["146"] = 209,["148"] = 213,["149"] = 214,["151"] = 218,["152"] = 220,["153"] = 222,["154"] = 224,["155"] = 225,["157"] = 229,["158"] = 230,["161"] = 236,["162"] = 237,["163"] = 238,["166"] = 244,["167"] = 245,["168"] = 246,["170"] = 252,["171"] = 253,["173"] = 261,["174"] = 262,["177"] = 268,["178"] = 269,["179"] = 270,["182"] = 275,["183"] = 276,["184"] = 277,["186"] = 284,["187"] = 285,["188"] = 286,["189"] = 286,["190"] = 286,["191"] = 286,["192"] = 286,["193"] = 286,["194"] = 286,["195"] = 286,["197"] = 290,["198"] = 186,["199"] = 33,["200"] = 33,["201"] = 33,["202"] = 33,["203"] = 33,["204"] = 33,["205"] = 33,["206"] = 33});
local ____exports = {}
local ____trigger = require("solar.solar-common.w3ts.handles.trigger")
local Trigger = ____trigger.Trigger
local ____AttributeUtil = require("solar.solar-common.util.system.AttributeUtil")
local AttributeUtil = ____AttributeUtil.default
local ____MathUtil = require("solar.solar-common.util.math.MathUtil")
local MathUtil = ____MathUtil.default
local ____SingletonUtil = require("solar.solar-common.util.lang.SingletonUtil")
local SingletonUtil = ____SingletonUtil.default
local ____DamageUtil = require("solar.solar-common.util.system.DamageUtil")
local DamageUtil = ____DamageUtil.default
local ____UnitStateUtil = require("solar.solar-common.util.unit.UnitStateUtil")
local UnitStateUtil = ____UnitStateUtil.default
____exports.default = __TS__Class()
local SolarDamageState = ____exports.default
SolarDamageState.name = "SolarDamageState"
function SolarDamageState.prototype.____constructor(self, useBaseAttribute)
    if useBaseAttribute == nil then
        useBaseAttribute = ____exports.default.config.useBaseAttribute
    end
    if SingletonUtil:notFirstTime(____exports.default) then
        print("不能重复new SolarDamageState()")
        return
    end
    ____exports.default.config.useBaseAttribute = useBaseAttribute
    if useBaseAttribute then
        ____exports.default:addEventHandler(____exports.default.calculateAttributeDamage)
    end
    isSolarDamageEnable = true
    local trigger = __TS__New(Trigger)
    trigger:registerAnyUnitDamagedEvent()
    trigger:addAction(self.action)
end
function SolarDamageState.addEventHandlerFirst(self, eventHandler)
    local ____exports_default_config_firstDamageEventHandlers_0 = ____exports.default.config.firstDamageEventHandlers
    ____exports_default_config_firstDamageEventHandlers_0[#____exports_default_config_firstDamageEventHandlers_0 + 1] = eventHandler
end
function SolarDamageState.addEventHandler(self, eventHandler)
    local ____exports_default_config_damageEventHandlers_1 = ____exports.default.config.damageEventHandlers
    ____exports_default_config_damageEventHandlers_1[#____exports_default_config_damageEventHandlers_1 + 1] = eventHandler
end
function SolarDamageState.addEventHandlerLast(self, eventHandler)
    local ____exports_default_config_lastDamageEventHandlers_2 = ____exports.default.config.lastDamageEventHandlers
    ____exports_default_config_lastDamageEventHandlers_2[#____exports_default_config_lastDamageEventHandlers_2 + 1] = eventHandler
end
function SolarDamageState.prototype.action()
    local eventDamage = GetEventDamage()
    if eventDamage <= 0 then
        return
    end
    local unit0 = GetTriggerUnit()
    local unit1 = GetEventDamageSource()
    if not IsHandle(unit1) then
        return
    end
    if IsUnitAlly(
        unit0,
        GetOwningPlayer(unit1)
    ) then
        return
    end
    local isAttack = 0 ~= EXGetEventDamageData(EVENT_DAMAGE_DATA_IS_ATTACK)
    if isBigAttributeMode and isAttack then
        local u1Attack = GetUnitState(unit1, UnitStateDamageMax)
        local u0Armor = GetUnitState(unit0, UnitStateArmor)
        eventDamage = u1Attack * (1 - MathUtil.armorReduction(u0Armor))
        EXSetEventDamage(eventDamage)
    end
    if not isSolarDamageEnable then
        return
    end
    local u0sa = AttributeUtil:getUnitAttribute(unit0)
    local u1sa = AttributeUtil:getUnitAttribute(unit1)
    local resultDamage = eventDamage
    if ____exports.default.config.useBaseAttribute and isAttack and u1sa ~= nil then
        if u1sa.def_pierce_p or u1sa.def_pierce then
            local armor = UnitStateUtil:getArmor(unit0) * (1 - (u1sa.def_pierce_p or 0))
            armor = armor - (u1sa.def_pierce or 0)
            local rd = MathUtil.armorReduction(armor)
            resultDamage = math.max(
                resultDamage,
                UnitStateUtil:getDamageMax(unit1)
            ) * (1 - rd)
        end
    end
    local event = {
        baseDamage = eventDamage,
        damageType = EXGetEventDamageData(EVENT_DAMAGE_DATA_DAMAGE_TYPE),
        weaponType = EXGetEventDamageData(EVENT_DAMAGE_DATA_WEAPON_TYPE),
        unit0 = unit0,
        unit1 = unit1,
        u0sa = u0sa,
        u1sa = u1sa,
        isAttack = isAttack,
        isRanged = 0 ~= EXGetEventDamageData(EVENT_DAMAGE_DATA_IS_RANGED),
        isPhysical = DamageUtil:isEventPhysicalDamageType(),
        isCritical = false,
        resultDamage = resultDamage
    }
    isSolarDamageEnable = false
    for ____, damageEventHandler in ipairs(____exports.default.config.firstDamageEventHandlers) do
        damageEventHandler(nil, event)
        if event.consumed then
            isSolarDamageEnable = true
            return
        end
    end
    for ____, damageEventHandler in ipairs(____exports.default.config.damageEventHandlers) do
        damageEventHandler(nil, event)
        if event.consumed then
            isSolarDamageEnable = true
            return
        end
    end
    for ____, damageEventHandler in ipairs(____exports.default.config.lastDamageEventHandlers) do
        damageEventHandler(nil, event)
        if event.consumed then
            isSolarDamageEnable = true
            return
        end
    end
    isSolarDamageEnable = true
    EXSetEventDamage(event.resultDamage)
end
function SolarDamageState.calculateAttributeDamage(self, event)
    local u0sa = event.u0sa
    local u1sa = event.u1sa
    if u0sa == nil and u1sa == nil then
        return
    end
    local newResultDamage = event.resultDamage
    if event.isAttack and (u1sa and u1sa.attack_damage_increased) then
        newResultDamage = newResultDamage * (u1sa.attack_damage_increased + 1)
    end
    if event.isPhysical then
        if ____exports.default.config.physical_damage_increased_attack or event.isAttack == false then
            newResultDamage = newResultDamage * ((u1sa and u1sa.physical_damage_increased or 0) + 1)
        end
        if GetRandomReal(0, 1) < (u1sa and u1sa.physical_critical_chance or 0) then
            newResultDamage = newResultDamage * ((u1sa and u1sa.physical_critical_damage or 0) + 1)
            event.isCritical = true
        end
        local physical_damage_pierce = (u1sa and u1sa.physical_damage_pierce or 0) - (u0sa and u0sa.physical_damage_reduction or 0)
        newResultDamage = newResultDamage * (1 + physical_damage_pierce)
    else
        newResultDamage = newResultDamage + (u1sa and u1sa.magic_power or 0)
        newResultDamage = newResultDamage * ((u1sa and u1sa.magic_damage_increased or 0) + 1)
        if GetRandomReal(0, 1) < (u1sa and u1sa.magic_critical_chance or 0) then
            newResultDamage = newResultDamage * ((u1sa and u1sa.magic_critical_damage or 0) + 1)
            event.isCritical = true
        end
        local magic_damage_pierce = (u1sa and u1sa.magic_damage_pierce or 0) - (u0sa and u0sa.magic_damage_reduction or 0)
        newResultDamage = newResultDamage * (1 + magic_damage_pierce)
    end
    --- 根据伤害类型 增伤
    local damage_increased_dt_x = "damage_increased_dt_" .. tostring(event.damageType)
    if u1sa and u1sa[damage_increased_dt_x] then
        newResultDamage = newResultDamage * (1 + u1sa[damage_increased_dt_x])
    end
    --- 根据武器类型 增伤
    local damage_increased_wt_x = "damage_increased_wt_" .. tostring(event.weaponType)
    if u1sa and u1sa[damage_increased_wt_x] then
        newResultDamage = newResultDamage * (1 + u1sa[damage_increased_wt_x])
    end
    if u1sa and u1sa.damage_increased then
        newResultDamage = newResultDamage * (1 + u1sa.damage_increased)
    end
    if u0sa and u0sa.damage_reduction then
        newResultDamage = newResultDamage * (1 - math.min(u0sa.damage_reduction, ____exports.default.config.damage_reduction_max))
    end
    --- 根据伤害类型 抗性 减伤
    local damage_reduction_dt_x = "damage_reduction_dt_" .. tostring(event.damageType)
    if u0sa and u0sa[damage_reduction_dt_x] then
        newResultDamage = newResultDamage * (1 - u0sa[damage_reduction_dt_x])
    end
    --- 根据武器类型 抗性 减伤
    local damage_reduction_wt_x = "damage_reduction_wt_" .. tostring(event.weaponType)
    if u0sa and u0sa[damage_reduction_wt_x] then
        newResultDamage = newResultDamage * (1 - u0sa[damage_reduction_wt_x])
    end
    if u1sa and u1sa.blood_sucking then
        local add_hp = newResultDamage * u1sa.blood_sucking
        SetUnitState(
            event.unit1,
            UNIT_STATE_LIFE,
            math.max(
                0,
                GetUnitState(event.unit1, UNIT_STATE_LIFE) + add_hp
            )
        )
    end
    event.resultDamage = newResultDamage
end
SolarDamageState.config = {
    useBaseAttribute = true,
    physical_damage_increased_attack = true,
    damage_reduction_max = 0.99,
    firstDamageEventHandlers = {},
    damageEventHandlers = {},
    lastDamageEventHandlers = {}
}
return ____exports
